File: backend/app/schemas/application.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.application import ApplicationStatus


class JobApplicationCreate(BaseModel):
    company: str = Field(min_length=1, max_length=255)
    position: str = Field(min_length=1, max_length=255)
    job_url: str | None = Field(default=None, max_length=2048)
    status: ApplicationStatus = ApplicationStatus.SAVED
    location: str | None = Field(default=None, max_length=255)
    salary: str | None = Field(default=None, max_length=100)
    applied_at: datetime | None = None
    deadline: datetime | None = None
    notes: str | None = None


class JobApplicationUpdate(BaseModel):
    company: str | None = Field(default=None, min_length=1, max_length=255)
    position: str | None = Field(default=None, min_length=1, max_length=255)
    job_url: str | None = Field(default=None, max_length=2048)
    status: ApplicationStatus | None = None
    location: str | None = Field(default=None, max_length=255)
    salary: str | None = Field(default=None, max_length=100)
    applied_at: datetime | None = None
    deadline: datetime | None = None
    notes: str | None = None


class JobApplicationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    company: str
    position: str
    job_url: str | None
    status: ApplicationStatus
    location: str | None
    salary: str | None
    applied_at: datetime | None
    deadline: datetime | None
    notes: str | None
    created_at: datetime
    updated_at: datetime


class JobApplicationListResponse(BaseModel):
    items: list[JobApplicationResponse]
    page: int
    page_size: int
    total: int
    pages: int

File: frontend/src/types/application.ts
export type ApplicationStatus =
  | 'SAVED'
  | 'APPLIED'
  | 'SCREENING'
  | 'INTERVIEW'
  | 'OFFER'
  | 'REJECTED'
  | 'WITHDRAWN'

export type Application = {
  id: number
  company: string
  position: string
  status: ApplicationStatus
  job_url: string | null
  location: string | null
  salary: string | null
  applied_at: string | null
  deadline: string | null
  notes: string | null
  created_at: string
  updated_at: string
}

export type ApplicationListResponse = {
  items: Application[]
  page: number
  page_size: number
  total: number
  pages: number
}

export type ApplicationSortBy =
  | 'created_at'
  | 'deadline'
  | 'company'
  | 'position'
  | 'status'

export type ApplicationSortOrder = 'asc' | 'desc'

export type ApplicationListParams = {
  search?: string
  status?: ApplicationStatus
  location?: string
  page?: number
  page_size?: number
  sort_by?: ApplicationSortBy
  sort_order?: ApplicationSortOrder
}

export type ApplicationCreateRequest = {
  company: string
  position: string
  job_url?: string | null
  status?: ApplicationStatus
  location?: string | null
  salary?: string | null
  applied_at?: string | null
  deadline?: string | null
  notes?: string | null
}

export type ApplicationUpdateRequest = {
  company?: string
  position?: string
  job_url?: string | null
  status?: ApplicationStatus
  location?: string | null
  salary?: string | null
  applied_at?: string | null
  deadline?: string | null
  notes?: string | null
}

export const APPLICATION_STATUS_ORDER: ApplicationStatus[] = [
  'SAVED',
  'APPLIED',
  'SCREENING',
  'INTERVIEW',
  'OFFER',
  'REJECTED',
  'WITHDRAWN',
]

export const APPLICATION_STATUS_LABELS: Record<ApplicationStatus, string> = {
  SAVED: 'Saved',
  APPLIED: 'Applied',
  SCREENING: 'Screening',
  INTERVIEW: 'Interview',
  OFFER: 'Offer',
  REJECTED: 'Rejected',
  WITHDRAWN: 'Withdrawn',
}
